Swift/4.0: Try to make it compiles and pass the tests

FredericRuaudel opened the pull request:
Hi!

   Here is a pull request to try to make the project work again with Swift 4.0. All the tests seems to pass following my changes so I hope I made no mistakes.

I'm not sure how you configured Carthage with submodules but it seems that my changes in Cartfile.private doesn't pull the last version of Nimble correctly (7.0.1) so maybe there is some things to tweak a little bit here to make the tests runs correctly. (On my computer, I had to fix a bug in AllPass.swift that should be fixed in the last Nimble version 🤔)

Hope it helps!

Thanks for this great parser framework!

Have a nice day,

Fred

FredericRuaudel commented:
Ok, I've finally found that I confused the version of Quick with the one of Nimble. I fixed that and manage to update my project with the last version. 

Now I have a bunch of modified file I'm not sure if I should commit them in the pull request or not:

```
        modified:   .gitmodules
	modified:   Carthage/Checkouts/Nimble (new commits)
	modified:   Carthage/Checkouts/Quick (new commits)
	modified:   Carthage/Checkouts/Result (new commits)
```

Tell me if it's needed.

inamiy commented:
Thanks for contribution!
Let me take some time to fix CI first before checking this PR.
BTW, adding git-submodules will be nicer :)

vknabel commented:
@inamiy @FredericRuaudel I could locally fix the SPM-Build error by applying the changes in https://gist.github.com/vknabel/a87ff19f18df0dac1f7e27867ad7957c/revisions.
Using the official Runes 4 at least allowed me to compile TryParsec.

FredericRuaudel commented:
> BTW, adding git-submodules will be nicer :)

You mean the 3 files I mentionned ?

```
        modified:   Carthage/Checkouts/Nimble (new commits)
	modified:   Carthage/Checkouts/Quick (new commits)
	modified:   Carthage/Checkouts/Result (new commits)
```

Or all 4 ?

I've hesitated because the .gitmodules file have replaced the absolute paths by relative ones and I'm not sure you are ok with that...

eg. :
```
-	url = https://github.com/antitypical/Result.git
+	url = ../../antitypical/Result
```

inamiy commented:
@FredericRuaudel 
Hm, in that case, it won't work 🙄  
Then, I will add some more commits to this PR later.

@vknabel 
I will check yours too!

inamiy commented on Sources/TryParsec/JSON.swift:
```diff
@@ -40,7 +40,7 @@ extension JSON: RawRepresentable
             case let .number(v): return v
             case let .bool(v):   return v
-            case let .null(v):   return v
+            case .null:          return ""
```
Fixed in 206a774.

inamiy commented on Sources/TryParsec/Parser+UnicodeScalarView.swift:
```diff
@@ -11,7 +11,7 @@ public func satisfy(_ predicate: @escaping (UnicodeScalar) -> Bool) -> Parser<St
     return Parser { input in
         if let (head, tail) = uncons(input), predicate(head) {
-            return .done(tail, head)
+            return .done(String(tail).unicodeScalars, head)
```
Interestingly, changing to `String.UnicodeScalarView(tail)` will hang up the test for some reason... 😕 
(Is there any reason that `Substring.UnicodeScalarView -> String.UnicodeScalarView` not working?)

Anyway, I will stick to this code for now.
